File: ios/ContractRunner/Sources/ContractRunner/PublicPersonalMediaContract.swift
import Foundation

public struct ContractMediaExternalIds: Equatable {
    public let tmdb: Int?
    public let imdb: String?
    public let tvdb: Int?

    public init(tmdb: Int?, imdb: String?, tvdb: Int?) {
        self.tmdb = tmdb
        self.imdb = imdb
        self.tvdb = tvdb
    }
}

public struct ContractUserItemData: Equatable {
    public let itemId: String?
    public let isFavorite: Bool?
    public let played: Bool?
    public let playCount: Int?
    public let playbackPositionSeconds: Double?
    public let runtimeSeconds: Double?
    public let playedPercentage: Double?
    public let lastPlayedDate: String?
    public let rating: Double?
    public let dismissedFromContinueWatching: Bool?

    public init(
        itemId: String? = nil,
        isFavorite: Bool? = nil,
        played: Bool? = nil,
        playCount: Int? = nil,
        playbackPositionSeconds: Double? = nil,
        runtimeSeconds: Double? = nil,
        playedPercentage: Double? = nil,
        lastPlayedDate: String? = nil,
        rating: Double? = nil,
        dismissedFromContinueWatching: Bool? = nil
    ) {
        self.itemId = itemId
        self.isFavorite = isFavorite
        self.played = played
        self.playCount = playCount
        self.playbackPositionSeconds = playbackPositionSeconds
        self.runtimeSeconds = runtimeSeconds
        self.playedPercentage = playedPercentage
        self.lastPlayedDate = lastPlayedDate
        self.rating = rating
        self.dismissedFromContinueWatching = dismissedFromContinueWatching
    }
}

public struct ContractMediaItem: Equatable {
    public let itemId: String
    public let mediaType: String
    public let title: String
    public let originalTitle: String?
    public let overview: String?
    public let posterUrl: String?
    public let backdropUrl: String?
    public let logoUrl: String?
    public let stillUrl: String?
    public let releaseDate: String?
    public let releaseYear: Int?
    public let rating: Double?
    public let genres: [String]
    public let runtimeMinutes: Int?
    public let status: String?
    public let certification: String?
    public let externalIds: ContractMediaExternalIds
    public let seasonNumber: Int?
    public let episodeNumber: Int?
    public let absoluteEpisodeNumber: Int?
    public let episodeTitle: String?
    public let airDate: String?
    public let tagline: String?
    public let seriesId: String?
    public let seriesName: String?
    public let seasonId: String?
    public let seasonName: String?
    public let userData: ContractUserItemData?

    public init(
        itemId: String,
        mediaType: String,
        title: String,
        originalTitle: String? = nil,
        overview: String? = nil,
        posterUrl: String? = nil,
        backdropUrl: String? = nil,
        logoUrl: String? = nil,
        stillUrl: String? = nil,
        releaseDate: String? = nil,
        releaseYear: Int? = nil,
        rating: Double? = nil,
        genres: [String] = [],
        runtimeMinutes: Int? = nil,
        status: String? = nil,
        certification: String? = nil,
        externalIds: ContractMediaExternalIds,
        seasonNumber: Int? = nil,
        episodeNumber: Int? = nil,
        absoluteEpisodeNumber: Int? = nil,
        episodeTitle: String? = nil,
        airDate: String? = nil,
        tagline: String? = nil,
        seriesId: String? = nil,
        seriesName: String? = nil,
        seasonId: String? = nil,
        seasonName: String? = nil,
        userData: ContractUserItemData? = nil
    ) {
        self.itemId = itemId
        self.mediaType = mediaType
        self.title = title
        self.originalTitle = originalTitle
        self.overview = overview
        self.posterUrl = posterUrl
        self.backdropUrl = backdropUrl
        self.logoUrl = logoUrl
        self.stillUrl = stillUrl
        self.releaseDate = releaseDate
        self.releaseYear = releaseYear
        self.rating = rating
        self.genres = genres
        self.runtimeMinutes = runtimeMinutes
        self.status = status
        self.certification = certification
        self.externalIds = externalIds
        self.seasonNumber = seasonNumber
        self.episodeNumber = episodeNumber
        self.absoluteEpisodeNumber = absoluteEpisodeNumber
        self.episodeTitle = episodeTitle
        self.airDate = airDate
        self.tagline = tagline
        self.seriesId = seriesId
        self.seriesName = seriesName
        self.seasonId = seasonId
        self.seasonName = seasonName
        self.userData = userData
    }
}

public struct ContractBaseItemDtoQueryResult: Equatable {
    public let items: [ContractMediaItem]
    public let startIndex: Int
    public let totalRecordCount: Int
    public let nextCursor: String?
    public let hasMore: Bool

    public init(items: [ContractMediaItem], startIndex: Int, totalRecordCount: Int, nextCursor: String?, hasMore: Bool) {
        self.items = items
        self.startIndex = startIndex
        self.totalRecordCount = totalRecordCount
        self.nextCursor = nextCursor
        self.hasMore = hasMore
    }
}

public struct ContractCalendarEnvelope: Equatable {
    public let profileId: String
    public let source: String
    public let kind: String?
    public let generatedAt: String
    public let items: [ContractMediaItem]

    public init(profileId: String, source: String, kind: String?, generatedAt: String, items: [ContractMediaItem]) {
        self.profileId = profileId
        self.source = source
        self.kind = kind
        self.generatedAt = generatedAt
        self.items = items
    }
}

public func normalizeBaseItemDtoQueryResult(payload: [String: Any]) -> ContractBaseItemDtoQueryResult? {
    guard hasExactKeys(payload, expected: ["Items", "StartIndex", "TotalRecordCount", "NextCursor", "HasMore"]),
          let itemValues = payload["Items"] as? [Any],
          let items = mapStrict(itemValues, transform: { value -> ContractMediaItem? in
              guard let object = value as? [String: Any] else { return nil }
              return parseMediaItem(object)
          }),
          let hasMore = payload["HasMore"] as? Bool else {
        return nil
    }
    let startIndex = nullableInt(payload, "StartIndex") ?? 0
    let totalRecordCount = nullableInt(payload, "TotalRecordCount") ?? items.count
    let nextCursor = optionalString(payload, "NextCursor")
    return ContractBaseItemDtoQueryResult(items: items, startIndex: startIndex, totalRecordCount: totalRecordCount, nextCursor: nextCursor, hasMore: hasMore)
}

public func normalizeCalendarEnvelope(payload: [String: Any]) -> ContractCalendarEnvelope? {
    guard let profileId = requiredString(payload, "profileId"),
          let source = requiredString(payload, "source"),
          let generatedAt = requiredString(payload, "generatedAt"),
          let itemValues = payload["items"] as? [Any],
          let items = mapStrict(itemValues, transform: { value -> ContractMediaItem? in
              guard let object = value as? [String: Any] else { return nil }
              return parseMediaItem(object)
          }) else {
        return nil
    }
    let kind = optionalString(payload, "kind")
    return ContractCalendarEnvelope(profileId: profileId, source: source, kind: kind, generatedAt: generatedAt, items: items)
}

private func parseMediaItem(_ payload: [String: Any]) -> ContractMediaItem? {
    guard let itemId = requiredString(payload, "Id"),
          let type = requiredString(payload, "Type"),
          let name = requiredString(payload, "Name") else {
        return nil
    }

    let imageTags = payload["ImageTags"] as? [String: Any]
    let providerIds = payload["ProviderIds"] as? [String: Any]
    let externalIds = ContractMediaExternalIds(
        tmdb: optionalString(providerIds ?? [:], "Tmdb").flatMap { Int($0) },
        imdb: optionalString(providerIds ?? [:], "Imdb"),
        tvdb: optionalString(providerIds ?? [:], "Tvdb").flatMap { Int($0) }
    )
    let userData = optionalObject(payload, "UserData").flatMap(parseContractUserItemData)
    let taglines = stringArray(payload["Taglines"] as? [Any] ?? [], key: "Taglines") ?? []

    return ContractMediaItem(
        itemId: itemId,
        mediaType: parseContractMediaItemType(type),
        title: name,
        originalTitle: optionalString(payload, "OriginalTitle"),
        overview: optionalString(payload, "Overview"),
        posterUrl: imageTagMedium(imageTags, "Primary"),
        backdropUrl: backdropMedium(imageTags),
        logoUrl: imageTagMedium(imageTags, "Logo"),
        stillUrl: imageTagMedium(imageTags, "Thumb"),
        releaseDate: optionalString(payload, "PremiereDate"),
        releaseYear: nullableInt(payload, "ProductionYear"),
        rating: nullableDouble(payload, "CommunityRating"),
        genres: stringArray(payload["Genres"] as? [Any] ?? [], key: "Genres") ?? [],
        runtimeMinutes: nullableDouble(payload, "RunTimeTicks").flatMap { $0 > 0 ? Int($0 / 600_000_000) : nil },
        status: optionalString(payload, "Status"),
        certification: optionalString(payload, "Certification"),
        externalIds: externalIds,
        seasonNumber: nullableInt(payload, "ParentIndexNumber"),
        episodeNumber: nullableInt(payload, "IndexNumber"),
        absoluteEpisodeNumber: nullableInt(payload, "AbsoluteIndexNumber"),
        episodeTitle: optionalString(payload, "EpisodeTitle"),
        airDate: optionalString(payload, "AirDate"),
        tagline: taglines.first,
        seriesId: optionalString(payload, "SeriesId"),
        seriesName: optionalString(payload, "SeriesName"),
        seasonId: optionalString(payload, "SeasonId"),
        seasonName: optionalString(payload, "SeasonName"),
        userData: userData
    )
}

private func parseContractMediaItemType(_ type: String) -> String {
    switch type.trimmingCharacters(in: .whitespacesAndNewlines) {
    case "Movie": return "movie"
    case "Series": return "show"
    case "Season": return "season"
    case "Episode": return "episode"
    case "Unknown": return "unknown"
    default: return "unknown"
    }
}

private func parseContractUserItemData(_ payload: [String: Any]) -> ContractUserItemData? {
    if payload.isEmpty { return nil }
    return ContractUserItemData(
        itemId: optionalString(payload, "ItemId"),
        isFavorite: payload["IsFavorite"] as? Bool,
        played: payload["Played"] as? Bool,
        playCount: (payload["PlayCount"] as? NSNumber)?.intValue,
        playbackPositionSeconds: doubleValue(payload["PlaybackPositionTicks"] as Any).map { $0 / 10_000_000 },
        runtimeSeconds: doubleValue(payload["RuntimeTicks"] as Any).map { $0 / 10_000_000 },
        playedPercentage: doubleValue(payload["PlayedPercentage"] as Any),
        lastPlayedDate: optionalString(payload, "LastPlayedDate"),
        rating: doubleValue(payload["Rating"] as Any),
        dismissedFromContinueWatching: payload["DismissedFromContinueWatching"] as? Bool
    )
}

private func imageTagMedium(_ tags: [String: Any]?, _ key: String) -> String? {
    guard let tag = tags?[key] else { return nil }
    if let string = tag as? String { return string }
    if let dict = tag as? [String: Any] { return optionalString(dict, "medium") }
    return nil
}

private func backdropMedium(_ tags: [String: Any]?) -> String? {
    guard let backdrops = tags?["Backdrop"] as? [Any], let first = backdrops.first else { return nil }
    if let string = first as? String { return string }
    if let dict = first as? [String: Any] { return optionalString(dict, "medium") }
    return nil
}

private func hasExactKeys(_ object: [String: Any], expected: Set<String>) -> Bool {
    return Set(object.keys) == expected
}

private func requiredString(_ object: [String: Any], _ key: String) -> String? {
    return object[key] as? String
}

private func optionalString(_ object: [String: Any], _ key: String) -> String? {
    guard let value = object[key] else {
        return nil
    }
    if value is NSNull {
        return nil
    }
    return value as? String
}

private func nullableInt(_ object: [String: Any], _ key: String) -> Int? {
    guard let value = object[key] else { return nil }
    if value is NSNull { return nil }
    return intValue(value)
}

private func intValue(_ value: Any) -> Int? {
    if let int = value as? Int {
        return int
    }
    if let number = value as? NSNumber {
        return number.intValue
    }
    if let string = value as? String {
        return Int(string.trimmingCharacters(in: .whitespacesAndNewlines))
    }
    return nil
}

private func doubleValue(_ value: Any) -> Double? {
    if let double = value as? Double {
        return double
    }
    if let number = value as? NSNumber {
        return number.doubleValue
    }
    if let string = value as? String {
        return Double(string.trimmingCharacters(in: .whitespacesAndNewlines))
    }
    return nil
}

private func mapStrict<T>(_ values: [Any], transform: (Any) -> T?) -> [T]? {
    var output: [T] = []
    for value in values {
        guard let transformed = transform(value) else {
            return nil
        }
        output.append(transformed)
    }
    return output
}

private func stringArray(_ values: [Any], key: String) -> [String]? {
    var output: [String] = []
    for value in values {
        if let string = value as? String {
            output.append(string)
        } else {
            return nil
        }
    }
    return output
}

private func optionalObject(_ object: [String: Any], _ key: String) -> [String: Any]? {
    guard let value = object[key] else { return nil }
    if value is NSNull { return nil }
    return value as? [String: Any]
}

private func nullableDouble(_ object: [String: Any], _ key: String) -> Double? {
    guard let value = object[key] else { return nil }
    if value is NSNull { return nil }
    return doubleValue(value)
}

File: android/core-domain/src/main/kotlin/com/crispy/tv/domain/media/PublicPersonalMediaContract.kt
package com.crispy.tv.domain.media

import com.crispy.tv.domain.MediaKey

data class ContractMediaExternalIds(
    val tmdb: Int?,
    val imdb: String?,
    val tvdb: Int?,
)

data class ContractMediaItem(
    val mediaKey: MediaKey,
    val mediaType: String,
    val title: String,
    val originalTitle: String?,
    val overview: String?,
    val posterUrl: String?,
    val backdropUrl: String?,
    val logoUrl: String?,
    val stillUrl: String?,
    val releaseDate: String?,
    val releaseYear: Int?,
    val rating: Double?,
    val genres: List<String>,
    val runtimeMinutes: Int?,
    val status: String?,
    val certification: String?,
    val externalIds: ContractMediaExternalIds,
    val seasonNumber: Int?,
    val episodeNumber: Int?,
    val absoluteEpisodeNumber: Int?,
    val episodeTitle: String?,
    val airDate: String?,
    val tagline: String?,
    val seriesId: String?,
    val seriesName: String?,
    val seasonId: String?,
    val seasonName: String?,
    val userData: ContractUserItemData?,
)

data class ContractUserItemData(
    val itemId: String?,
    val isFavorite: Boolean?,
    val played: Boolean?,
    val playCount: Int?,
    val playbackPositionSeconds: Double?,
    val runtimeSeconds: Double?,
    val playedPercentage: Double?,
    val lastPlayedDate: String?,
    val rating: Double?,
    val dismissedFromContinueWatching: Boolean?,
)

data class ContractBaseItemDtoQueryResult(
    val items: List<ContractMediaItem>,
    val startIndex: Int,
    val totalRecordCount: Int,
    val nextCursor: String?,
    val hasMore: Boolean,
)

data class ContractCalendarEnvelope(
    val profileId: String,
    val source: String,
    val kind: String?,
    val generatedAt: String,
    val items: List<ContractMediaItem>,
)

fun normalizeBaseItemDtoQueryResult(payload: Map<String, Any?>): ContractBaseItemDtoQueryResult? {
    if (!payload.hasExactKeys(setOf("Items", "StartIndex", "TotalRecordCount", "NextCursor", "HasMore"))) return null
    val items = payload.requiredList("Items")?.mapStrict { value ->
        (value as? Map<*, *>)?.toStringAnyMap()?.let(::parseMediaItem)
    } ?: return null
    val startIndex = payload.nullableInt("StartIndex") ?: 0
    val totalRecordCount = payload.nullableInt("TotalRecordCount") ?: items.size
    val nextCursor = payload.nullableString("NextCursor")
    val hasMore = payload.requiredBoolean("HasMore") ?: return null
    return ContractBaseItemDtoQueryResult(items, startIndex, totalRecordCount, nextCursor, hasMore)
}

fun normalizeCalendarEnvelope(payload: Map<String, Any?>): ContractCalendarEnvelope? {
    if (!payload.hasRequiredKeys(setOf("profileId", "source", "generatedAt", "items"))) return null
    val profileId = payload.requiredString("profileId") ?: return null
    val source = payload.requiredString("source") ?: return null
    val generatedAt = payload.requiredString("generatedAt") ?: return null
    val items = payload.requiredList("items")?.mapStrict { value ->
        (value as? Map<*, *>)?.toStringAnyMap()?.let(::parseMediaItem)
    } ?: return null
    val kind = payload.nullableString("kind")
    return ContractCalendarEnvelope(profileId, source, kind, generatedAt, items)
}

private fun parseMediaItem(payload: Map<String, Any?>): ContractMediaItem? {
    if (!payload.hasRequiredKeys(setOf("Id", "Type", "Name"))) return null
    val mediaKey = payload.requiredString("Id") ?: return null
    val type = payload.requiredString("Type") ?: return null
    val name = payload.requiredString("Name") ?: return null
    val imageTags = payload.nullableObject("ImageTags")
    return ContractMediaItem(
        mediaKey = MediaKey(mediaKey),
        mediaType = parseContractMediaItemType(type),
        title = name,
        originalTitle = payload.nullableString("OriginalTitle"),
        overview = payload.nullableString("Overview"),
        posterUrl = imageTags?.imageTagMedium("Primary"),
        backdropUrl = imageTags?.backdropMedium(),
        logoUrl = imageTags?.imageTagMedium("Logo"),
        stillUrl = imageTags?.imageTagMedium("Thumb"),
        releaseDate = payload.nullableString("PremiereDate"),
        releaseYear = payload.nullableInt("ProductionYear"),
        rating = payload.nullableNumber("CommunityRating"),
        genres = payload.requiredStringList("Genres") ?: return null,
        runtimeMinutes = payload.nullableNumber("RunTimeTicks")?.toLong()?.let { if (it > 0L) (it / 600_000_000L).toInt() else null },
        status = payload.nullableString("Status"),
        certification = payload.nullableString("Certification"),
        externalIds = payload.nullableObject("ProviderIds")?.let(::parseProviderIds) ?: ContractMediaExternalIds(null, null, null),
        seasonNumber = payload.nullableInt("ParentIndexNumber"),
        episodeNumber = payload.nullableInt("IndexNumber"),
        absoluteEpisodeNumber = payload.nullableInt("AbsoluteIndexNumber"),
        episodeTitle = payload.nullableString("EpisodeTitle"),
        airDate = payload.nullableString("AirDate"),
        tagline = payload.requiredStringList("Taglines")?.firstOrNull(),
        seriesId = payload.nullableString("SeriesId"),
        seriesName = payload.nullableString("SeriesName"),
        seasonId = payload.nullableString("SeasonId"),
        seasonName = payload.nullableString("SeasonName"),
        userData = payload.nullableObject("UserData")?.let(::parseContractUserItemData),
    )
}

private fun parseProviderIds(payload: Map<String, Any?>): ContractMediaExternalIds {
    return ContractMediaExternalIds(
        tmdb = payload.nullableString("Tmdb")?.toIntOrNull(),
        imdb = payload.nullableString("Imdb"),
        tvdb = payload.nullableString("Tvdb")?.toIntOrNull(),
    )
}

private fun parseContractUserItemData(payload: Map<String, Any?>): ContractUserItemData? {
    if (payload.isEmpty()) return null
    return ContractUserItemData(
        itemId = payload.nullableString("ItemId"),
        isFavorite = payload["IsFavorite"] as? Boolean,
        played = payload["Played"] as? Boolean,
        playCount = (payload["PlayCount"] as? Number)?.toInt(),
        playbackPositionSeconds = (payload["PlaybackPositionTicks"] as? Number)?.toDouble()?.let { it / 10_000_000.0 },
        runtimeSeconds = (payload["RuntimeTicks"] as? Number)?.toDouble()?.let { it / 10_000_000.0 },
        playedPercentage = (payload["PlayedPercentage"] as? Number)?.toDouble(),
        lastPlayedDate = payload.nullableString("LastPlayedDate"),
        rating = (payload["Rating"] as? Number)?.toDouble(),
        dismissedFromContinueWatching = payload["DismissedFromContinueWatching"] as? Boolean,
    )
}

private fun parseContractMediaItemType(type: String): String {
    return when (type.trim()) {
        "Movie" -> "movie"
        "Series" -> "show"
        "Season" -> "season"
        "Episode" -> "episode"
        "Unknown" -> "unknown"
        else -> "unknown"
    }
}

private fun Map<String, Any?>.imageTagMedium(key: String): String? {
    val tag = this[key]
    return when (tag) {
        is String -> tag
        is Map<*, *> -> tag.toStringAnyMap()?.nullableString("medium")
        else -> null
    }
}

private fun Map<String, Any?>.backdropMedium(): String? {
    val backdrops = this["Backdrop"] as? List<*> ?: return null
    val first = backdrops.firstOrNull() ?: return null
    return when (first) {
        is String -> first.trim().ifBlank { null }
        is Map<*, *> -> first.toStringAnyMap()?.nullableString("medium")
        else -> null
    }
}

private fun Map<String, Any?>.hasExactKeys(expected: Set<String>): Boolean = keys == expected

private fun Map<String, Any?>.hasRequiredKeys(required: Set<String>): Boolean = required.all { containsKey(it) }

private fun Map<String, Any?>.requiredString(key: String): String? = this[key] as? String

private fun Map<String, Any?>.nullableString(key: String): String? {
    if (!containsKey(key)) return null
    val value = this[key]
    return when (value) {
        null -> null
        is String -> value
        else -> return null
    }
}

private fun Map<String, Any?>.requiredBoolean(key: String): Boolean? = this[key] as? Boolean

private fun Map<String, Any?>.requiredNumber(key: String): Double? = (this[key] as? Number)?.toDouble()

private fun Map<String, Any?>.nullableNumber(key: String): Double? {
    if (!containsKey(key)) return null
    val value = this[key]
    return when (value) {
        null -> null
        is Number -> value.toDouble()
        else -> return null
    }
}

private fun Map<String, Any?>.nullableInt(key: String): Int? {
    if (!containsKey(key)) return null
    val value = this[key]
    return when (value) {
        null -> null
        is Int -> value
        is Long -> value.toInt()
        is Number -> value.toDouble().takeIf { it % 1.0 == 0.0 }?.toInt()
        else -> return null
    }
}

private fun Map<String, Any?>.requiredObject(key: String): Map<String, Any?>? = (this[key] as? Map<*, *>)?.toStringAnyMap()

private fun Map<String, Any?>.nullableObject(key: String): Map<String, Any?>? {
    if (!containsKey(key)) return null
    val value = this[key] ?: return null
    return (value as? Map<*, *>)?.toStringAnyMap()
}

private fun Map<String, Any?>.requiredList(key: String): List<Any?>? = this[key] as? List<Any?>

private fun Map<String, Any?>.requiredStringList(key: String): List<String>? {
    val values = this[key] as? List<*> ?: return null
    return values.mapStrict { value -> value as? String }
}

private fun <T, R> List<T>.mapStrict(transform: (T) -> R?): List<R>? {
    val results = ArrayList<R>(size)
    for (value in this) {
        results += transform(value) ?: return null
    }
    return results
}

private fun Map<*, *>.toStringAnyMap(): Map<String, Any?>? {
    val results = LinkedHashMap<String, Any?>(size)
    for ((key, value) in this) {
        val stringKey = key as? String ?: return null
        results[stringKey] = value
    }
    return results
}
